src/librustc_typeck/check/mod.rs

```diff
@@ -3855,10 +3855,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
               let ctxt = BreakableCtxt {
                   // cannot use break with a value from a while loop
                   coerce: None,
-                  may_break: true,
+                  may_break: false,  // Will get updated if/when we find a `break`.
               };
 
-              self.with_breakable_ctxt(expr.id, ctxt, || {
+              let (ctxt, ()) = self.with_breakable_ctxt(expr.id, ctxt, || {
                   self.check_expr_has_type_or_error(&cond, tcx.types.bool);
                   let cond_diverging = self.diverges.get();
```

Comment:
So what's confusing me here: there is somehow an implicit `break` should the condition be false (this is why `may_break` used to be considered `true`), and I can't see where that is accounted for.

For example, does this code accept:

```rust
let x: ! = {
  while false { return; };
};
```

( Do we have a test for this? It annoys me that this question is so hard to answer =) )

Reply:
Traditionally we have considered `while` loops as *always* potentially breaking — but it seems like your change could allow us to handle cases like this one (for example)

```rust
'a: loop {
    let x: ! = {
        while (continue 'a) { };
    };
}
```